### PR TITLE
NvFP4 quantized LM head support

### PR DESCRIPTION
**Pull request by ynankani**
## Overview

Add support for quantized LM head.

1) Create output_scale tensor with weight tying check.
2) Pass output_s in build_lora_mm, consistent with other scale tensors like wq_s.

## Additional information

 vLLM is also supporting Quantized LM head with this [PR ](https://github.com/vllm-project/vllm/pull/42124)
 
## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Yes, for redundant work and testing the change

### CODE REVIEW

File: src/llama-model.cpp [hunk 1]
```diff
@@ -1396,6 +1396,14 @@ bool llama_model_base::load_tensors(llama_model_loader & ml) {
         }
     }
 
+    const bool is_output_tied = output && output == tok_embd;
+    if(!is_output_tied && !output_s && output){
+        output_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "scale"), {1}, TENSOR_NOT_REQUIRED);
+    }
+    if(!is_output_tied && !output_in_s && output){
+        output_in_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "input_scale"), {1}, TENSOR_NOT_REQUIRED);
+    }
```
Comment (CISC) [inline src/llama-model.cpp]: This should be moved into the `create tensors for the weights` block above, and both into the same `if (!is_output_tied) {}`.

**Review by CISC — APPROVED** [review]

File: src/llama-model.cpp [hunk 2]
```diff
@@ -1394,8 +1394,17 @@ bool llama_model_base::load_tensors(llama_model_loader & ml) {
...
+        const bool is_output_tied = output && output == tok_embd;
+        if(!is_output_tied){
+            if(!output_s && output){
+                output_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "scale"), {1}, TENSOR_NOT_REQUIRED);
+            }
+            if(!output_in_s && output){
+                output_in_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "input_scale"), {1}, TENSOR_NOT_REQUIRED);
+            }
+        }
+    }
```
Comment (gaugarg-nv) [inline src/llama-model.cpp]:
```suggestion
        if(output && output != tok_embd){
            if(!output_s) {
                output_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "scale"), {1}, TENSOR_NOT_REQUIRED);
            }
            if(!output_in_s) {
                output_in_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "input_scale"), {1}, TENSOR_NOT_REQUIRED);
            }
        }
    }
```

Comment (gaugarg-nv) [reply src/llama-model.cpp]: Do we really need to add this check for weight tying? Since we don't support `tok_embd` as NVFP4, should we add an assert to ensure weight tying isn't enabled when the LM head is in NVFP4?

**Comment by CISC** [comment]
Looks like this somehow broke `phimoe`?

File: src/llama-model.cpp (hunk 2 above)
Comment (CISC) [reply src/llama-model.cpp]: I suppose a `GGML_ASSERT` would be in order here as output scale tensors and tied embedding will fail either way.

**Comment by ynankani** [comment]
> like this somehow broke `phimoe`?

Issue here is the pointer for tok_embd and output are different and create tensor on test path creates tensor by default if dimensions are given ignoring the TENSOR_NOT_REQUIRED. So there is a count mismatch .
1) one option was to compare the name instead of pointer for output and tok_embd
2) Check dtype for LM head if NvFP4 then only create output_s  tensor called at least on test path.
3) Remove the weight tying check just add assert when weight tying and lm head in NvFP4

File: src/llama-model.cpp [hunk 3]
```diff
@@ -1394,10 +1394,25 @@ bool llama_model_base::load_tensors(llama_model_loader & ml) {
...
+            }
+        }
     }
-
     ml.done_getting_tensors();
 
+    if (output && tok_embd &&
+            std::string(output->name) == std::string(tok_embd->name) &&
+            output->type == GGML_TYPE_NVFP4) {
+        throw std::runtime_error("NVFP4 quantized LM head is not supported with tied embeddings");
```
Comment (gaugarg-nv) [inline src/llama-model.cpp]: I think a `GGML_ASSERT` will be better.

File: src/llama-model.cpp [hunk 4]
```diff
@@ -1394,10 +1394,25 @@ bool llama_model_base::load_tensors(llama_model_loader & ml) {
...
+            if(!output_in_s){
+                output_in_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "input_scale"), {1}, TENSOR_NOT_REQUIRED);
+            }
+        }
     }
-
     ml.done_getting_tensors();
 
+    if (output && tok_embd &&
+            std::string(output->name) == std::string(tok_embd->name) &&
```
Comment (gaugarg-nv) [inline src/llama-model.cpp]:
```suggestion
            strcmp(output->name, tok_embd->name) == 0 &&
```

Nit.

File: src/llama-model.cpp [hunk 5]
```diff
@@ -1394,10 +1394,25 @@ bool llama_model_base::load_tensors(llama_model_loader & ml) {
...
+            // weight scale
+            if(!output_s){
+                output_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "scale"), {1}, TENSOR_NOT_REQUIRED);
+            }
+            // input scale
+            if(!output_in_s){
+                output_in_s = create_tensor(tn(LLM_TENSOR_OUTPUT, "input_scale"), {1}, TENSOR_NOT_REQUIRED);
+            }
+        }
     }
```
Comment (gaugarg-nv) [inline src/llama-model.cpp]: Nit: There are some formatting issues here.

GGML follows `if(...) {`. Note the space before `{`.

**Review by gaugarg-nv — APPROVED** [review]

**Comment by ynankani** [comment]
Test cases are failing due to Vulkan tanh shader doesn't handle near zero values. If i revert #10723  i don't see the test failures.
Option 1
```
use: data_d[i] = D_TYPE(tanh(float(data_a[i]))); 
instead of : data_d[i] = D_TYPE(1. - 2. / (exp(2.*data_a[i]) + 1.));
```

Other solution can be to use 

```
float x = data_d[i]
if (abs(x) < 2.4414e-4) {
    // tanh(x) ≈ x - x^3/3 + 2x^5/15 , no exp, no cancellation
    float x2 = x * x;
    result = x * (1.0 - x2 * (1.0/3.0 - x2 * (2.0/15.0)));
} else {
    result = 1.0 - 2.0 / (exp(2.0 * x) + 1.0);
}
```

@0cc4m Please let me know if the #10723 PR is still needed for Vulkan

**Comment by 0cc4m** [comment]
I don't understand this change yet. How does it cause different use of `tanh`?

**Comment by ynankani** [comment]
> I don't understand this change yet. How does it cause different use of `tanh`?

This change adds output_s tensor for LM head in NvFp4. In the test path as there doesn't a gguf file it creates a tensor with the given dimension and a value is set using Normal distribution for the tensor. The output_s is then used in mat-mul. The tanh has different behavior for Vulkan and CPU for near zero values which causes high  NMSE  for Vulkan path. 
1. logits = output @ x                          (normal magnitudes)
2. logits *= output_s                            (N(0, 0.01), so ~0.001-0.1)
3. logits /= f_final_logit_softcapping (30)      (~3e-5 to 3e-3)
4. logits = tanh(logits)                        <= Need to handle near zero value
5. logits *= 30                                  (back to ~0.001-0.1)

**Comment by gaugarg-nv** [comment]
IMO, the current change to add the tensors only when the type is NVFP4 should be fine. Running the CI again to see if it fixes the Vulkan failure.

**Review by gaugarg-nv — APPROVED** [review]
